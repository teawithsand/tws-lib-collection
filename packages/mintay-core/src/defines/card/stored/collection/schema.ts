import { z } from "zod"

const StoredSdelkaCollectionDataV1Schema = z.object({
	globalId: z.string(),
	title: z.string(),
	description: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCollectionDataSchema = z.discriminatedUnion(
	"version",
	[
		z.object({
			version: z.literal(1),
			data: StoredSdelkaCollectionDataV1Schema,
		}),
	],
)

export type StoredSdelkaCollectionData = z.infer<
	typeof StoredSdelkaCollectionDataSchema
>
