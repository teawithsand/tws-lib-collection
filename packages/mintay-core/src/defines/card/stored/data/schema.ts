import { z } from "zod"

const StoredSdelkaCardDataV1Schema = z.object({
	globalId: z.string(),
	content: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCardDataSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardDataV1Schema,
	}),
])

export type StoredSdelkaCardData = z.infer<typeof StoredSdelkaCardDataSchema>
