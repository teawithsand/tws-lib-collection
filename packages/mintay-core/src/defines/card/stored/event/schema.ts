import { z } from "zod"
import { StoredSdelkaCardEventV1Schema } from "./schemaV1"

export const StoredSdelkaCardEventSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardEventV1Schema,
	}),
])

export type StoredSdelkaCardEvent = z.infer<typeof StoredSdelkaCardEventSchema>
