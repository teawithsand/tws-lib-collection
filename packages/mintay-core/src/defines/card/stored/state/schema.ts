import { z } from "zod"
import { StoredSdelkaCardStateV1Schema } from "./schemaV1"

export const StoredSdelkaCardStateSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardStateV1Schema,
	}),
])

export type StoredSdelkaCardState = z.infer<typeof StoredSdelkaCardStateSchema>
