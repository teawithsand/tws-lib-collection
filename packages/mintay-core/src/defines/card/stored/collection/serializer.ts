import { SdelkaCollectionData } from "../../sdelkaCollectionData"
import { StoredSdelkaCollectionData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCollectionDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCollectionData,
	): StoredSdelkaCollectionData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				title: data.title,
				description: data.description,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCollectionData,
	): SdelkaCollectionData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				title: d.title,
				description: d.description,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
