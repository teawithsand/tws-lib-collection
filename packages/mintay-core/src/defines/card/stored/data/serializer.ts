import { SdelkaCardData } from "../../sdelkaCardData"
import { StoredSdelkaCardData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCardDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCardData,
	): StoredSdelkaCardData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				content: data.content,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCardData,
	): SdelkaCardData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				content: d.content,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
