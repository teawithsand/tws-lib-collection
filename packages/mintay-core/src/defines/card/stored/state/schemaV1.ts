import { z } from "zod"

export enum StoredSdelkaCardQueueV1 {
	NEW = 0,
	LEARNING = 1,
	LEARNED = 2,
	RELEARNING = 3,
}

const StoredSdelkaCardStateFSRSSchema = z.object({
	dueTimestamp: z.number(),
	stability: z.number(),
	difficulty: z.number(),
	elapsedDays: z.number(),
	scheduledDays: z.number(),
	reps: z.number(),
	lapses: z.number(),
	state: z.nativeEnum(StoredSdelkaCardQueueV1),
	lastReviewTimestamp: z.number().nullable(),
})

export const StoredSdelkaCardStateV1Schema = z.object({
	fsrs: StoredSdelkaCardStateFSRSSchema,
})
export type StoredSdelkaCardStateV1 = z.infer<
	typeof StoredSdelkaCardStateV1Schema
>
