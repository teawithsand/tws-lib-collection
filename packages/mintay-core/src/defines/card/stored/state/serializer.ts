import { SdelkaCardState } from "../../sdelkaCardState"
import { SdelkaCardQueue } from "../../sdelkaQueue"
import { StoredSdelkaCardState } from "./schema"
import { StoredSdelkaCardQueueV1 } from "./schemaV1"

export class SdelkaCardStateSerializer {
	private constructor() {}

	private static readonly toStoredQueue = (
		queue: SdelkaCardQueue,
	): StoredSdelkaCardQueueV1 => {
		switch (queue) {
			case SdelkaCardQueue.NEW:
				return StoredSdelkaCardQueueV1.NEW
			case SdelkaCardQueue.LEARNING:
				return StoredSdelkaCardQueueV1.LEARNING
			case SdelkaCardQueue.LEARNED:
				return StoredSdelkaCardQueueV1.LEARNED
			case SdelkaCardQueue.RELEARNING:
				return StoredSdelkaCardQueueV1.RELEARNING
		}
		const _exhaustiveCheck: never = queue
		return _exhaustiveCheck
	}

	private static readonly fromStoredQueue = (
		queue: StoredSdelkaCardQueueV1,
	): SdelkaCardQueue => {
		switch (queue) {
			case StoredSdelkaCardQueueV1.NEW:
				return SdelkaCardQueue.NEW
			case StoredSdelkaCardQueueV1.LEARNING:
				return SdelkaCardQueue.LEARNING
			case StoredSdelkaCardQueueV1.LEARNED:
				return SdelkaCardQueue.LEARNED
			case StoredSdelkaCardQueueV1.RELEARNING:
				return SdelkaCardQueue.RELEARNING
		}
		const _exhaustiveCheck: never = queue
		return _exhaustiveCheck
	}

	public static readonly serialize = (
		state: SdelkaCardState,
	): StoredSdelkaCardState => {
		return {
			version: 1,
			data: {
				fsrs: {
					dueTimestamp: state.fsrs.dueTimestamp,
					stability: state.fsrs.stability,
					difficulty: state.fsrs.difficulty,
					elapsedDays: state.fsrs.elapsedDays,
					scheduledDays: state.fsrs.scheduledDays,
					reps: state.fsrs.reps,
					lapses: state.fsrs.lapses,
					state: SdelkaCardStateSerializer.toStoredQueue(
						state.fsrs.state,
					),
					lastReviewTimestamp: state.fsrs.lastReviewTimestamp,
				},
			},
		}
	}

	public static readonly deserialize = (
		stored: StoredSdelkaCardState,
	): SdelkaCardState => {
		if (stored.version === 1) {
			const data = stored.data
			return {
				fsrs: {
					dueTimestamp: data.fsrs.dueTimestamp,
					stability: data.fsrs.stability,
					difficulty: data.fsrs.difficulty,
					elapsedDays: data.fsrs.elapsedDays,
					scheduledDays: data.fsrs.scheduledDays,
					reps: data.fsrs.reps,
					lapses: data.fsrs.lapses,
					state: SdelkaCardStateSerializer.fromStoredQueue(
						data.fsrs.state,
					),
					lastReviewTimestamp: data.fsrs.lastReviewTimestamp,
				},
			}
		}
		const _exhaustiveCheck: never = stored.version
		return _exhaustiveCheck
	}
}
