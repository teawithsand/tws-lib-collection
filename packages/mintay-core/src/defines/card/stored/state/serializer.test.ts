import { describe, expect, test } from "vitest"
import { SdelkaCardState } from "../../sdelkaCardState"
import { SdelkaCardQueue } from "../../sdelkaQueue"
import { StoredSdelkaCardState } from "./schema"
import { StoredSdelkaCardQueueV1 } from "./schemaV1"
import { SdelkaCardStateSerializer } from "./serializer"

describe("SdelkaCardStateSerializer", () => {
	const exampleStateBase: Omit<SdelkaCardState, "fsrs"> & {
		fsrs: Omit<SdelkaCardState["fsrs"], "state">
	} = {
		fsrs: {
			dueTimestamp: 1234567890,
			stability: 0.8,
			difficulty: 0.5,
			elapsedDays: 10,
			scheduledDays: 5,
			reps: 20,
			lapses: 2,
			lastReviewTimestamp: 1234560000,
		},
	}

	const statesToTest = [
		SdelkaCardQueue.NEW,
		SdelkaCardQueue.LEARNING,
		SdelkaCardQueue.LEARNED,
		SdelkaCardQueue.RELEARNING,
	]

	statesToTest.forEach((state) => {
		test(`serialize and deserialize with state ${StoredSdelkaCardQueueV1[state]} (${state})`, () => {
			const exampleState = {
				...exampleStateBase,
				fsrs: {
					...exampleStateBase.fsrs,
					state,
				},
			}
			const serialized = SdelkaCardStateSerializer.serialize(exampleState)
			expect(serialized.data.fsrs.state).toBe(state)
			const deserialized =
				SdelkaCardStateSerializer.deserialize(serialized)
			expect(deserialized.fsrs.state).toBe(exampleState.fsrs.state)
			expect(deserialized).toEqual(exampleState)
		})
	})

	const exampleState: SdelkaCardState = {
		fsrs: {
			dueTimestamp: 1234567890,
			stability: 0.8,
			difficulty: 0.5,
			elapsedDays: 10,
			scheduledDays: 5,
			reps: 20,
			lapses: 2,
			state: SdelkaCardQueue.LEARNING,
			lastReviewTimestamp: 1234560000,
		},
	}

	test("serialize returns correct StoredSdelkaCardState with version 1", () => {
		const serialized = SdelkaCardStateSerializer.serialize(exampleState)
		expect(serialized.version).toBe(1)
		expect(serialized.data.fsrs.dueTimestamp).toBe(
			exampleState.fsrs.dueTimestamp,
		)
		expect(serialized.data.fsrs.stability).toBe(exampleState.fsrs.stability)
		expect(serialized.data.fsrs.difficulty).toBe(
			exampleState.fsrs.difficulty,
		)
		expect(serialized.data.fsrs.elapsedDays).toBe(
			exampleState.fsrs.elapsedDays,
		)
		expect(serialized.data.fsrs.scheduledDays).toBe(
			exampleState.fsrs.scheduledDays,
		)
		expect(serialized.data.fsrs.reps).toBe(exampleState.fsrs.reps)
		expect(serialized.data.fsrs.lapses).toBe(exampleState.fsrs.lapses)
		expect(serialized.data.fsrs.state).toBe(1)
		expect(serialized.data.fsrs.lastReviewTimestamp).toBe(
			exampleState.fsrs.lastReviewTimestamp,
		)
	})

	test("serialize and then deserialize should not change the state", () => {
		const serialized = SdelkaCardStateSerializer.serialize(exampleState)
		const deserialized = SdelkaCardStateSerializer.deserialize(serialized)
		expect(deserialized).toEqual(exampleState)
	})

	test("deserialize returns correct SdelkaCardState from StoredSdelkaCardState", () => {
		const stored: StoredSdelkaCardState = {
			version: 1,
			data: {
				fsrs: {
					dueTimestamp: 1234567890,
					stability: 0.8,
					difficulty: 0.5,
					elapsedDays: 10,
					scheduledDays: 5,
					reps: 20,
					lapses: 2,
					state: 1,
					lastReviewTimestamp: 1234560000,
				},
			},
		}
		const deserialized = SdelkaCardStateSerializer.deserialize(stored)
		expect(deserialized.fsrs.dueTimestamp).toBe(
			stored.data.fsrs.dueTimestamp,
		)
		expect(deserialized.fsrs.stability).toBe(stored.data.fsrs.stability)
		expect(deserialized.fsrs.difficulty).toBe(stored.data.fsrs.difficulty)
		expect(deserialized.fsrs.elapsedDays).toBe(stored.data.fsrs.elapsedDays)
		expect(deserialized.fsrs.scheduledDays).toBe(
			stored.data.fsrs.scheduledDays,
		)
		expect(deserialized.fsrs.reps).toBe(stored.data.fsrs.reps)
		expect(deserialized.fsrs.lapses).toBe(stored.data.fsrs.lapses)
		expect(deserialized.fsrs.state).toBe(SdelkaCardQueue.LEARNING)
		expect(deserialized.fsrs.lastReviewTimestamp).toBe(
			stored.data.fsrs.lastReviewTimestamp,
		)
	})

	test("deserialize and then serialize should not change the stored state", () => {
		const stored: StoredSdelkaCardState = {
			version: 1,
			data: {
				fsrs: {
					dueTimestamp: 1234567890,
					stability: 0.8,
					difficulty: 0.5,
					elapsedDays: 10,
					scheduledDays: 5,
					reps: 20,
					lapses: 2,
					state: 1,
					lastReviewTimestamp: 1234560000,
				},
			},
		}
		const deserialized = SdelkaCardStateSerializer.deserialize(stored)
		const reStored = SdelkaCardStateSerializer.serialize(deserialized)
		expect(reStored).toEqual(stored)
	})

	test("serialize snapshot", () => {
		const serialized = SdelkaCardStateSerializer.serialize(exampleState)
		expect(serialized).toMatchSnapshot()
	})

	test("deserialize snapshot", () => {
		const stored: StoredSdelkaCardState = {
			version: 1,
			data: {
				fsrs: {
					dueTimestamp: 1234567890,
					stability: 0.8,
					difficulty: 0.5,
					elapsedDays: 10,
					scheduledDays: 5,
					reps: 20,
					lapses: 2,
					state: 1,
					lastReviewTimestamp: 1234560000,
				},
			},
		}
		const deserialized = SdelkaCardStateSerializer.deserialize(stored)
		expect(deserialized).toMatchSnapshot()
	})
})
