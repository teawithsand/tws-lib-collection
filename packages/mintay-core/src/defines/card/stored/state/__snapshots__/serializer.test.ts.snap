// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SdelkaCardStateSerializer > deserialize snapshot 1`] = `
{
  "fsrs": {
    "difficulty": 0.5,
    "dueTimestamp": 1234567890,
    "elapsedDays": 10,
    "lapses": 2,
    "lastReviewTimestamp": 1234560000,
    "reps": 20,
    "scheduledDays": 5,
    "stability": 0.8,
    "state": 1,
  },
}
`;

exports[`SdelkaCardStateSerializer > serialize snapshot 1`] = `
{
  "data": {
    "fsrs": {
      "difficulty": 0.5,
      "dueTimestamp": 1234567890,
      "elapsedDays": 10,
      "lapses": 2,
      "lastReviewTimestamp": 1234560000,
      "reps": 20,
      "scheduledDays": 5,
      "stability": 0.8,
      "state": 1,
    },
  },
  "version": 1,
}
`;
